rand_distr/src/utils.rs

```diff
@@ -91,33 +111,33 @@ impl Float for f64 {
 /// `Ag(z)` is an infinite series with coefficients that can be calculated
 /// ahead of time - we use just the first 6 terms, which is good enough
 /// for most purposes.
-pub(crate) fn log_gamma(x: f64) -> f64 {
+pub(crate) fn log_gamma<N: Float>(x: N) -> N {
```

Comment:
This is not an optimal `f32` implementation and I'm not sure it's even an acceptable one (e.g. the `a` parameter below rounds to `1`). I would suggest just using `f64` arithmetic and converting at function start/end, though there's very little point to supporting `Poisson<f32>` in this case (only really for the type deduction).

Additionally, this function prototype is okay for internal use but not if we wished to expose `log_gamma` since we could not add correct implementations for other float types without specialization. A trait would be better, though clunky since we only have one method.

Statrs's Gamma function implementations are likely better.

Reply:
It essentially evaluates a polynomial, so I think it should be fine, even if the coefficients have less precision. I thinks it is more likely that too many coefficients are used for the target precision.

Reply:
> though there's very little point to supporting Poisson<f32> in this case (only really for the type deduction).

It might help for sampling from the uniform distribution.

Reply:
How does this help with the uniform distribution?

Reply:
We are doing some [rejection sampling for the Poisson distribution](https://github.com/vks/rand/blob/84d89a7c77206d92eccff1b70c42f7074d6eea3b/rand_distr/src/poisson.rs#L115), sampling `f32` instead of `f64` in the loop might help.

Reply:
Ah, you mean speed up usage of uniform, not help implement uniform. Yes, and when we get a specialisation of `log_gamma` it may speed things up for any(?) situation where `f32` is faster than `f64`.